Source/Research.h

```diff
@@ -0,0 +1,71 @@
+// Research.h
+// Classes for modifying Research structures
+#ifndef _RESEARCH_H_
+#define _RESEARCH_H_
+
+enum LabType
+{
+	ltBasic			= 1,
+	ltStandard		= 2,
+	ltAdvanced		= 3
+};
+
+// "CATEGORY" values (for both items and upgrades to these items)
+enum TechCategory
+{
+	tcFree			= 0, // 0 = Free technologies (and unavailable technologies)
+	tcBasic			= 1, // 1 = Basic labratory sciences
+	tcDefense		= 2, // 2 = Defenses (GP upgrade, walls, and efficiency engineering)
+	tcPower			= 3, // 3 = Power
+	tcVehicles		= 4, // 4 = Vehicles (which ones can be built, speed upgrades, armour upgrades)
+	tcFood			= 5, // 5 = Food
+	tcMetals		= 6, // 6 = Metals gathering
+	tcWeapons		= 7, // 7 = Weapons
+	tcSpace			= 8, // 8 = Space (spaceport, observatory, launch vehicle, skydock)
+	tcMorale		= 9, // 9 = Population (happiness)
+	tcDisaster		= 10,// 10 = Disaster warning (and defense)
+	tcPopulation	= 11,// 11 = Population (health, growth)
+	tcSpaceship		= 12 // 12 = Spaceship module
+};
+
+class TechInfo
```

Comment:
Why the split between `TechInfo` and `OP2TechInfo`? Could probably have a simpler design with just one struct, with `public` fields, which are directly accessed. It would be easy enough to cast a pointer or reference to a struct to get access to the data. Would also allow for easily copying the data, should you want your own copy of it.

Reply:
I did this to mimic the style of the other files. If you look at PlayerEx.h/cpp, you can see there is PlayerEx (the access layer) and OP2Player (the memory hacked struct). Same thing is going on here, where "OP2" in OP2TechInfo is signaling the struct is a "memory hack". Whether or not it makes sense to do it this way wasn't taken into consideration, but I like the idea of an access layer that can't blow up in your face if you use it wrong.

Reply:
Ahh, you'll have to excuse me for being a little ignorant about pre-existing practice here. Back in the day, I had my own little library to do this sort of thing, which eventually grew into what is now OP2Internal. It just did pointer casting of structs and classes, rather than writing a bunch of accessor methods on handle classes.